lib/prelude.dx

```diff
@@ -406,6 +398,123 @@ def finEq (n:Int) ?-> : Eq (Fin n) = MkEq \x y. ordinal x == ordinal y
 def finOrd (n:Int) ?-> : Ord (Fin n) =
   MkOrd finEq (\x y. ordinal x > ordinal y) (\x y. ordinal x < ordinal y)
 
+'## Raw pointer operations
+
+RawPtr : Type = %Word8Ptr
+
+data Ptr a:Type = MkPtr RawPtr
+
+-- Is there a better way to select the right instance for `storageSize`??
+data TypeVehicle a:Type = MkTypeVehicle
+def typeVehicle (a:Type) : TypeVehicle a = MkTypeVehicle
+
+interface Storable a:Type where
+  store : Ptr a -> a -> {State World} Unit
+  load  : Ptr a ->      {State World} a
+  storageSize : TypeVehicle a -> Int
+
+-- TODO: we can't inline these into the instance definitions until we change
+-- type inference to push types down into record constructors or allow `def` in
+-- instance definitions.
+def word8Store ((MkPtr ptr): Ptr Word8) (x:Word8) : {State World} Unit  = %ptrStore ptr x
+def word8Load  ((MkPtr ptr): Ptr Word8)           : {State World} Word8 = %ptrLoad  ptr
+
+instance word8Storable : Storable Word8 where
+  store = word8Store
+  load  = word8Load
+  storageSize = const 1
+
+-- TODO: there's a bug preventing us inlining these definitions into the instance
+def int32Store ((MkPtr ptr): Ptr Int32) (x:Int32) : {State World} Unit  =
+  %ptrStore (internalCast %Int32Ptr ptr) x
+def int32Load  ((MkPtr ptr): Ptr Int32) : {State World} Int32 =
+  %ptrLoad (internalCast %Int32Ptr ptr)
+
+instance int32Storable : Storable Int32 where
+  store = int32Store
+  load  = int32Load
+  storageSize = const 4
+
+def unpackPairPtr (_:Storable a) ?=> (_:Storable b) ?=>
+      (pairPtr: Ptr (a & b)) : (Ptr a & Ptr b) =
+  (MkPtr rawPtrX) = pairPtr
+  rawPtrY = %ptrOffset rawPtrX (storageSize (typeVehicle a))
```

Comment:
This is... a little sketchy because of the alignment requirements. Can we try to avoid tuples in FFI? It's higly non-standard.

Reply:
Great point. I never think about alignment. I'll remove the tuple Storable instance.